internal/server/polling/server.go

```diff
@@ -0,0 +1,107 @@
+package polling
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/weaveworks/tf-controller/internal/git/provider"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const DefaultPollingInterval = time.Second * 30
+
+type Server struct {
+	log             logr.Logger
+	clusterClient   client.Client
+	configMapRef    client.ObjectKey
+	pollingInterval time.Duration
+}
+
+func New(options ...Option) (*Server, error) {
```

Comment:
It can be, based on the provider. For example for github only Token is mandatory (i think), but other providers may require different type of token structure that needs a different function. For now only token would be a mandatory argument as a single string. Other providers may require different token that has 2 parts, this way we don't have to change the function signature in the future.

Reply:
This makes me think perhaps the API is not quite the right shape: it sounds like only some permutations of options will give you a valid object (on API design, I always go back to http://sweng.the-davies.net/Home/rustys-api-design-manifesto). But, I lack a constructive suggestion for improving it right now, so I'll drop this bone :-)